XForm/XForm/Functions/Coalesce.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Collections.Generic;
+using XForm.Data;
+using XForm.Query;
+
+namespace XForm.Functions
+{
+    public class Coalesce : IFunctionBuilder
+    {
+        public Type ReturnType => null;
+
+        public string Usage => "Coalesce({Col|Func|Const}, ...)";
+
+        public string Name => "Coalesce";
+
+        public IXColumn Build(IXTable source, XDatabaseContext context)
+        {
+            return new CoalesceTransformFunction(source, context);
+        }
+    }
+
+    public class CoalesceTransformFunction : IXColumn
+    {
+        private List<IXColumn> _columns = new List<IXColumn>();
+        private ICoalescer _coalescer;
+
+        public CoalesceTransformFunction(IXTable source, XDatabaseContext context)
+        {
+            while (context.Parser.HasAnotherArgument)
+            {
+                IXColumn column = context.Parser.NextColumn(source, context);
+                _columns.Add(column);
+                if (System.String.IsNullOrEmpty(column.ColumnDetails.Name)) throw new ArgumentException($"Column {_columns.Count} passed to 'Column' wasn't assigned a name. Use 'AS [Name]' to assign names to every column selected.");
+            }
+
+            if (!AllTypesMatch(_columns))
+                throw new UsageException("Coalesce requires all column types to match");
+
+            _coalescer = (ICoalescer)Allocator.ConstructGenericOf(typeof(Coalescer<>), _columns[0].ColumnDetails.Type);
+
+            ColumnDetails = new ColumnDetails($"{_columns[0].ColumnDetails.Name}.Coalesce", _columns[0].ColumnDetails.Type);
+            IndicesType = _columns[0].IndicesType;
+        }
+
+        public ColumnDetails ColumnDetails { get; private set; }
+
+        public Type IndicesType { get; private set; }
+
+        public Func<XArray> CurrentGetter()
+        {
+            Func<ArraySelector, XArray>[] getters = new Func<ArraySelector, XArray>[_columns.Count];
+
+            for (int index = 0; index < _columns.Count; index++)
+            {
+                int i = index;
+                getters[i] = (selector) => _columns[i].CurrentGetter()();
```

Comment:
I think the way this is written will keep getting the CurrentGetter each time it's called.
Any logic inside the lambda will run every time.

I think you want:
// Cache the getters
Func<XArray>[] getters = _columns.Select((col) => col.CurrentGetter()).ToArray();
Func<ArraySelector, XArray>[] selectedGetters = getters.Select((getter) => (selector) => getter()).ToArray();

Subtly different. Basically, make sure you don't have two pairs of parens back to back.
 +